File: headpin/katello-headpin/katello/lib/task.rb
#
# Copyright 2011 Red Hat, Inc.
#
# This software is licensed to you under the GNU General Public
# License as published by the Free Software Foundation; either version
# 2 of the License (GPLv2) or (at your option) any later version.
# There is NO WARRANTY for this software, express or implied,
# including the implied warranties of MERCHANTABILITY,
# NON-INFRINGEMENT, or FITNESS FOR A PARTICULAR PURPOSE. You should
# have received a copy of GPLv2 along with this software; if not, see
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.

class Task
  attr_reader :name, :status, :priority, :action, :timestamp

  def initialize opts
    @name      = opts[:name]
    @status    = opts[:status]
    @priority  = opts[:priority] || 0
    @action    = opts[:action]
    update_ts
  end

  def status=s
    if Queue::STATUS.include?(s)
      update_ts
      @status = s
    else
      raise "invalid STATE #{s}"
    end
  end

  def to_s
    "#{name}\t #{priority}\t #{status}\t #{action}"
  end

  private
  def update_ts
    @timestamp = Time.now
  end

  # sort based on priority
  def <=> other
    self.priority <=> other.priority
  end
end

File: cli/src/katello/client/core/task.py
#
# Katello Organization actions
# Copyright 2013 Red Hat, Inc.
#
# This software is licensed to you under the GNU General Public License,
# version 2 (GPLv2). There is NO WARRANTY for this software, express or
# implied, including the implied warranties of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
# along with this software; if not, see
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
#
# Red Hat trademarks are not licensed under GPLv2. No permission is
# granted to use or replicate Red Hat trademarks that are incorporated
# in this software or its documentation.
#

import os

from katello.client.api.task_status import TaskStatusAPI
from katello.client.core.base import BaseAction, Command
from katello.client.api.utils import ApiDataError

# base task action ----------------------------------------------------------------

class TaskAction(BaseAction):

    def __init__(self):
        super(TaskAction, self).__init__()
        self.api = TaskStatusAPI()

# task actions --------------------------------------------------------------------

class Status(TaskAction):

    description = _("get a task's status")

    def setup_parser(self, parser):
        parser.add_option("--uuid", dest='uuid',
                          help=_("task uuid eg: c9668eda-096b-445d-b96d (required)"))

    def check_options(self, validator):
        validator.require(('uuid'))

    def run(self):
        uuid = self.get_option('uuid')

        task = self.api.status(uuid)
        if task is None:
            raise ApiDataError(_("Could not find task [ %s ].") % uuid)

        self.printer.add_column('uuid', _("UUID"))
        self.printer.add_column('state', _("State"))
        self.printer.add_column('progress', _("Progress"))
        self.printer.add_column('start_time', _("Start Time"))
        self.printer.add_column('finish_time', ("Finish Time"))
        self.printer.set_header(_("Task Status"))
        self.printer.print_item(task)
        return os.EX_OK

# task command --------------------------------------------------------------------

class Task(Command):

    description = _('commands for retrieving task information')
